File: client-android/lib/src/main/kotlin/stasis/client_android/lib/telemetry/analytics/AnalyticsPersistence.kt
package stasis.client_android.lib.telemetry.analytics

import stasis.client_android.lib.utils.Try
import java.time.Instant

interface AnalyticsPersistence {
    fun cache(entry: AnalyticsEntry)
    suspend fun transmit(entry: AnalyticsEntry): Try<Unit>
    suspend fun restore(): Try<AnalyticsEntry?>

    val lastCached: Instant
    val lastTransmitted: Instant
}

File: layers/src/main/scala/stasis/layers/telemetry/analytics/AnalyticsPersistence.scala
package stasis.layers.telemetry.analytics

import java.time.Instant

import scala.concurrent.Future

import org.apache.pekko.Done

trait AnalyticsPersistence {
  type Self

  def cache(entry: AnalyticsEntry): Unit
  def transmit(entry: AnalyticsEntry): Future[Done]
  def restore(): Future[Option[AnalyticsEntry]]

  def lastCached: Instant
  def lastTransmitted: Instant

  def withClientProvider(provider: AnalyticsClient.Provider): Self
}
